Sources/PubNub/EventEngine/Subscribe/SubscribeTransition.swift

```diff
@@ -0,0 +1,325 @@
+      return [.cancel(id: Invocation.ID.ReceiveMessages)]
+    case is Subscribe.ReceiveReconnectingState:
+      return [.cancel(id: Invocation.ID.ReceiveReconnect)]
+    default:
+      return []
+    }
+  }
+  
+  func transition(
+    from state: State,
+    event: Event
+  ) -> TransitionResult<State, Invocation> {
+    var results: TransitionResult<State, Invocation>
+    
+    switch event {
+    case .reconnect:
+      switch state {
+      case let state as any CursorState:
+        results = setReceivingState(from: state, cursor: state.cursor)
+      default:
+        results = setHandshakingState(from: state)
+      }
+    case .handshakeSucceess(let cursor):
+      results = setReceivingState(from: state, cursor: cursor)
+    case .handshakeFailure(let error):
+      results = setHandshakeReconnectingState(from: state, error: error)
+    case .handshakeReconnectSuccess(let cursor):
+      results = setReceivingState(from: state, cursor: cursor)
+    case .handshakeReconnectFailure(let error):
+      results = setHandshakeReconnectingState(from: state, error: error)
+    case .handshakeReconnectGiveUp(let error):
+      results = setHandshakeFailedState(from: state, error: error)
+    case .receiveSuccess(let cursor, let messages):
+      results = setReceivingState(from: state, cursor: cursor, messages: messages)
+    case .receiveFailure(let error):
+      results = setReceiveReconnectingState(from: state, error: error)
+    case .receiveReconnectSuccess(let cursor, let messages):
+      results = setReceivingState(from: state, cursor: cursor, messages: messages)
+    case .receiveReconnectFailure(let error):
+      results = setReceiveReconnectingState(from: state, error: error)
+    case .receiveReconnectGiveUp(let error):
+      results = setReceiveFailedState(from: state, error: error)
+    case .subscriptionChanged(let channels, let groups):
+      results = onSubscriptionChanged(from: state, newChannels: channels, newGroups: groups)
+    case .subscriptionRestored(let channels, let groups, let cursor):
+      results = onSubscriptionRestored(from: state, newChannels: channels, newGroups: groups, cursor: cursor)
+    case .disconnect:
+      results = setStoppedState(from: state)
+    }
+    
+    return TransitionResult(
+      state: results.state,
+      invocations: willExit(from: state) + results.invocations
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func onSubscriptionChanged(
+    from state: State,
+    newChannels: [String],
+    newGroups: [String]
+  ) -> TransitionResult<State, Invocation> {
+    return TransitionResult(
+      state: Subscribe.HandshakingState(
+        input: SubscribeInput(
+          channels: newChannels,
+          groups: newGroups
+        )
+      ),
+      invocations: [
+        .managed(
+          invocation: .handshakeRequest(
+            channels: newChannels,
+            groups: newGroups
+          )
+        )
+      ]
+    )
```

Comment:
`SUBSCRIPTION_CHANGED` event also possible on `Receiving` and it's corresponding state variants and should be targeted to `Receiving` state.